### PR TITLE
Allows shots to be passed to transforms at runtime

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The number of shots is currently a reserved runtime argument when calling QNodes. This PR enables this for batch transforms as well.

**Description of the Change:**

- When a batch transform is called with `shots` as keyword argument, this is intercepted and used as a runtime argument when calling `qml.execute()`.

- In addition, a small bug was found and fixed at the same time, whereby the number of shots was not returned to the original value after execution.

**Benefits:** Transforms can also be dynamically called with different shot values.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/interfaces/batch/__init__.py [hunk 1]
```diff
@@ -66,8 +66,7 @@ def set_shots(device, shots):
             device.shots = shots
         yield
     finally:
-        if device.shots != original_shots:
-            device.shots = original_shots
```
Comment (josh146) [inline pennylane/interfaces/batch/__init__.py:70]: @glassnotes this is the bug affecting the notebook - `!=` isn't safe enough, better to simply always revert the shots.

File: pennylane/interfaces/batch/__init__.py [hunk 2]
```diff
@@ -59,15 +59,18 @@ def set_shots(device, shots):
     >>> set_shots(dev, shots=100)(lambda: dev.shots)()
     100
     """
+    if shots == device.shots:
+        yield
+        return
```
Comment (josh146) [inline pennylane/interfaces/batch/__init__.py:64]: Now, if `shots == device.shots`, we yield and do nothing!

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -280,6 +280,23 @@
           return ([[c, 0.0, 2 * x], [-c, 0.0, 0.0]],)
   ```
 
+* Shots can now be passed as a runtime argument to batch transforms, similarly
```
Comment (mariaschuld) [inline doc/releases/changelog-dev.md:283]:
```suggestion
* Shots can now be passed as a runtime argument to transforms that execute circuits in batches, similarly
```

Is this what you mean? Even I don't know what a *batch transform* is, so better make it explicit? :)

File: tests/gradients/test_gradient_transform.py [hunk 4]
```diff
@@ -347,6 +347,28 @@ def circuit(weights):
...
+        x = 0.543
+
+        # the gradient function can be called with different shot values
+        grad_fn = qml.gradients.param_shift(circuit)
+        assert grad_fn(x).shape == (1, 1)
+        assert grad_fn(x, shots=[(1, 1000)]).shape == (1000, 1)
+
+        # the original QNode is unaffected
+        assert circuit(x).shape == tuple()
+        assert circuit(x, shots=1000).shape == tuple()
```
Comment (mariaschuld) [inline tests/gradients/test_gradient_transform.py:370]: What happens if a circuit uses the shots argument? This was always a tricky case, but shall we define intended behaviour with a test?

File: doc/releases/changelog-dev.md (hunk 3 above)
Comment (mariaschuld) [reply doc/releases/changelog-dev.md:283]: Oh wait, you mean the actual `batch_transform` function? Maybe use the function name then, so people can look it up?

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -280,6 +280,23 @@
...
+  to QNodes.
+  [(#1707)](https://github.com/PennyLaneAI/pennylane/pull/1707)
+
+  ```pycon
+  >>> dev = qml.device("default.qubit", wires=1, shots=1000)
+  >>> @qml.beta.qnode(dev)
+  ... def circuit(x):
+  ...     qml.RX(x, wires=0)
+  ...     return qml.expval(qml.PauliZ(0))
+  >>> grad_fn = qml.gradients.param_shift(circuit)
```
Comment (mariaschuld) [inline doc/releases/changelog-dev.md:293]: I was intuitively looking for `batch_transform`, but I guess `param_shift` uses `batch_transform`? Can we mention how the example relates?

File: pennylane/gradients/gradient_transform.py [hunk 6]
```diff
@@ -161,6 +161,7 @@ def jacobian_wrapper(*args, **kwargs):
             if not hybrid:
                 return qjac
 
+            kwargs.pop("shots", False)
```
Comment (mariaschuld) [inline pennylane/gradients/gradient_transform.py:164]: Why is this needed?

File: pennylane/interfaces/batch/__init__.py (hunk 2 above)
Comment (mariaschuld) [reply pennylane/interfaces/batch/__init__.py:64]: Ah good idea!

**Review by mariaschuld — APPROVED** [review]
I still have a very bad feeling about this syntax, but since we have it, it makes only sense to add it to the batch transforms! 

Approving, but I would strongly recommend to improve the changelog :)

File: doc/releases/changelog-dev.md (hunk 3 above)
Comment (josh146) [reply doc/releases/changelog-dev.md:283]: Oh I meant the first one :) Thanks, your suggestion is clearer!

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -280,6 +280,23 @@
           return ([[c, 0.0, 2 * x], [-c, 0.0, 0.0]],)
   ```
 
+* Shots can now be passed as a runtime argument to batch transforms, similarly
+  to QNodes.
+  [(#1707)](https://github.com/PennyLaneAI/pennylane/pull/1707)
```
Comment (josh146) [inline doc/releases/changelog-dev.md:285]:
```suggestion
  [(#1707)](https://github.com/PennyLaneAI/pennylane/pull/1707)
  
  An example of such a transform are the gradient transforms in the
  `qml.gradients` module. As a result, we can now call gradient transforms
  (such as `qml.gradients.param_shift`) and set the number of shots at runtime.
```

File: pennylane/gradients/gradient_transform.py (hunk 6 above)
Comment (josh146) [reply pennylane/gradients/gradient_transform.py:164]: Ah, your comment just made me think of a better way, allowing this to be removed!

File: tests/gradients/test_gradient_transform.py (hunk 4 above)
Comment (josh146) [reply tests/gradients/test_gradient_transform.py:370]: We currently raise a deprecation warning in the QNode, let me check if it also occurs here, or if one should be added manually 🤔 

https://github.com/PennyLaneAI/pennylane/blob/8a14365e206771874dcfb48ef6d68d51de9118ba/pennylane/qnode.py#L182
